### PR TITLE
Remove legacy acl tokens

### PR DESCRIPTION
**Pull request by skpratt**
### Description
This change removes all uses of legacy ACL tokens from consul.
Subsequent PRs will remove all legacy policies (https://github.com/hashicorp/consul/pull/15922), and update language to clarify that the AccessorID is now the primary identifier for management (https://github.com/hashicorp/consul/pull/15922).

### Testing & Reproduction steps
Tested locally and on EKS. Bootstrapped ACLs for 3 server cluster with 2 clients, tested normal UI and token-related CLI and API commands.

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: website/content/api-docs/acl/index.mdx [hunk 1]
```diff
@@ -201,6 +201,7 @@ $ curl \
 - `LastErrorMessage` - The last error message produced at the time of `LastError`.
   An empty string indicates that no sync has resulted in an error.
 
+<<<<<<< HEAD
```
Comment (cthain) [inline website/content/api-docs/acl/index.mdx:204]: Not sure if this is intentional.. looks like an in-progress git merge. There are some other ones below as well.

File: website/content/commands/acl/policy/create.mdx [hunk 2]
```diff
@@ -9,10 +9,7 @@ Command: `consul acl policy create`
 
 Corresponding HTTP API Endpoint: [\[PUT\] /v1/acl/policy](/consul/api-docs/acl/policies#create-a-policy)
 
-The `acl policy create` command creates new policies. The policies rules can either be set explicitly or the
-`-from-token` parameter may be used to load the rules from a legacy ACL token. When loading
-the rules from an existing legacy ACL token, the rules get translated from the legacy syntax
-to the new syntax.
+The `acl policy create` command creates new policies.
 
 Both the `-rules` and `-from-token` parameter values allow loading the value
```
Comment (cthain) [inline website/content/commands/acl/policy/create.mdx:14]: I think the `-from-token` ref needs to be removed.

**Review by cthain — COMMENTED** [review]
Finally got through it all. Nice work @skpratt! I left my comments about removing certain functions from the API client in so that you could comment and provide context for others on why they are still present instead of being removed.

File: api/acl.go [hunk 3]
```diff
@@ -90,7 +90,7 @@ type ACLTokenListEntry struct {
 	ExpirationTime    *time.Time `json:",omitempty"`
 	CreateTime        time.Time
 	Hash              []byte
-	Legacy            bool
+	Legacy            bool `json:",omitempty"` // DEPRECATED
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/acl.go]: updated JSON tag and discussed api stability concerns offline

File: .changelog/15947.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+```release-note:deprecation
```
[the remark this answers is not in the document]
Comment (skpratt) [reply .changelog/15947.txt]: good call, done

File: api/acl.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ type ACLToken struct {
 	Hash              []byte        `json:",omitempty"`
 
 	// DEPRECATED (ACL-Legacy-Compat)
-	// Rules will only be present for legacy tokens returned via the new APIs
+	// Rules are an artifact of legacy tokens deprecated in Consul 1.4
 	Rules string `json:",omitempty"`
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/acl.go]: updated JSON tag and discussed api stability concerns offline

File: agent/consul/leader.go [hunk 6]
```diff
@@ -553,96 +550,6 @@ func (s *Server) initializeACLs(ctx context.Context) error {
...
-func (s *Server) startACLUpgrade(ctx context.Context) {
-	if s.config.PrimaryDatacenter != s.config.Datacenter {
-		// token upgrades should only run in the primary
-		return
-	}
-
-	s.leaderRoutineManager.Start(ctx, aclUpgradeRoutineName, s.legacyACLTokenUpgrade)
-}
-
 func (s *Server) stopACLUpgrade() {
```
[the remark this answers is not in the document]
Comment (skpratt) [reply agent/consul/leader.go]: thank you for the catch! done

File: agent/consul/state/acl.go [hunk 7]
```diff
@@ -451,23 +447,12 @@ func aclTokenSetTxn(tx WriteTxn, idx uint64, token *structs.ACLToken, opts ACLTo
...
-	// DEPRECATED (ACL-Legacy-Compat)
-	if token.Rules != "" {
-		// When we update a legacy acl token we may have to correct old HCL to
-		// prevent the propagation of older syntax into the state store and
-		// into in-memory representations.
-		correctedRules := structs.SanitizeLegacyACLTokenRules(token.Rules)
-		if correctedRules != "" {
-			token.Rules = correctedRules
-		}
+	if token.AccessorID == "" {
```
[the remark this answers is not in the document]
Comment (skpratt) [reply agent/consul/state/acl.go:450]: agreed, better to surface this error as early as possible

File: api/acl.go [hunk 8]
```diff
@@ -1042,58 +1042,19 @@ func (a *ACL) PolicyList(q *QueryOptions) ([]*ACLPolicyListEntry, *QueryMeta, er
...
-	parseQueryMeta(resp, qm)
-	qm.RequestTime = rtt
-
-	ruleBytes, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return "", fmt.Errorf("Failed to read translated rule body: %v", err)
-	}
-
-	return string(ruleBytes), nil
+	return "", fmt.Errorf("Legacy ACL rules were deprecated in Consul 1.4")
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/acl.go:1048]: discussed api stability concerns offline

File: api/acl.go [hunk 9]
```diff
@@ -1042,58 +1042,19 @@ func (a *ACL) PolicyList(q *QueryOptions) ([]*ACLPolicyListEntry, *QueryMeta, er
...
-	parseQueryMeta(resp, qm)
-	qm.RequestTime = rtt
-
-	ruleBytes, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return "", fmt.Errorf("Failed to read translated rule body: %v", err)
-	}
-
-	return string(ruleBytes), nil
+	return "", fmt.Errorf("Legacy ACL tokens and rules were deprecated in Consul 1.4")
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/acl.go:1057]: discussed api stability concerns offline

File: api/agent.go [hunk 10]
```diff
@@ -1268,35 +1268,35 @@ func (a *Agent) monitor(loglevel string, logJSON bool, stopCh <-chan struct{}, q
...
 
 // UpdateACLReplicationToken updates the agent's "acl_replication_token". See
-// updateToken for more details.
+// updateToken for more details. Deprecated in Consul 1.4.
 //
 // DEPRECATED (ACL-Legacy-Compat) - Prefer UpdateReplicationACLToken for v1.4.3 and above
 func (a *Agent) UpdateACLReplicationToken(token string, q *WriteOptions) (*WriteMeta, error) {
-	return a.updateToken("acl_replication_token", token, q)
+	return nil, fmt.Errorf("Legacy ACL Tokens were deprecated in Consul 1.4")
 }
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/agent.go:1300]: discussed api stability concerns offline

File: api/operator_license.go [hunk 11]
```diff
@@ -30,6 +30,9 @@ type License struct {
 	// no longer be used in any capacity
 	TerminationTime time.Time `json:"termination_time"`
 
+	// Whether the license will ignore termination
+	IgnoreTermination bool `json:"ignore_termination"`
```
[the remark this answers is not in the document]
Comment (skpratt) [reply api/operator_license.go]: accidental bad patch artifact - moved to separate PR

File: website/content/api-docs/acl/index.mdx (hunk 1 above)
Comment (skpratt) [reply website/content/api-docs/acl/index.mdx]: bad patch artifact - apologies for sloppiness

File: website/content/commands/acl/policy/create.mdx (hunk 2 above)
Comment (skpratt) [reply website/content/commands/acl/policy/create.mdx]: thank you for this catch

File: website/content/api-docs/acl/index.mdx (hunk 1 above)
Comment (cthain) [reply website/content/api-docs/acl/index.mdx]: No worries!

**Review by cthain — APPROVED** [review]
:+1:

**Pull request merged**